### PR TITLE
Building Spring Boot 2.0 projects with Gradle

### PR DESCRIPTION
**Pull request by guylabs**
See https://github.com/gradle/guides/issues/186

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -4,13 +4,24 @@ plugins {
...
 }
 
+repositories {
+    maven {
+        url = uri("https://repo.gradle.org/gradle/libs")
+    }
+}
+
+dependencies {
+    implementation("org.gradle:sample-check:0.1.0")
```
Comment (eriwen) [inline build.gradle.kts]: I think we want `testImplementation`, and the newest version of sample-check, which is 0.4.2.

File: contents/index.adoc [hunk 2]
```diff
@@ -1,68 +1,235 @@
 = Building Spring Boot 2 Applications with Gradle
 
-TODO: Preamble goes here
-
-== What you'll build
-
-TODO
+This blog post shows how to build a new Gradle project for the lately released
```
Comment (eriwen) [inline contents/index.adoc]: "blog post" => "guide"

File: contents/index.adoc [hunk 3]
```diff
@@ -1,68 +1,235 @@
...
+Spring Boot 2.0. First we show some new and noteworthy features of Spring Boot
```
Comment (eriwen) [inline contents/index.adoc]: I'm not super picky about this, but we try to keep a convention of breaking lines on sentences, and not so much on line length. I recognize how annoying such a change would be unless you're feeling frisky with your regex, so please consider this optional.

File: contents/index.adoc [hunk 4]
```diff
@@ -1,68 +1,235 @@
...
+and its Gradle plugin. Next we’ll setup the Gradle project, apply the Spring
+Boot plugin, use the new Gradle BOM support to define the dependencies and
+create an example project to show an integration with Gradle Build Scans.
```
Comment (eriwen) [inline contents/index.adoc]: ❓ Link to build scans?

File: contents/index.adoc [hunk 5]
```diff
@@ -1,68 +1,235 @@
...
 
-TODO: Preamble goes here
-
-== What you'll build
-
-TODO
+This blog post shows how to build a new Gradle project for the lately released
+Spring Boot 2.0. First we show some new and noteworthy features of Spring Boot
+and its Gradle plugin. Next we’ll setup the Gradle project, apply the Spring
+Boot plugin, use the new Gradle BOM support to define the dependencies and
```
Comment (eriwen) [inline contents/index.adoc]: "new" might be omitted here, as soon enough it won't be "new" anymore.

File: contents/index.adoc [hunk 6]
```diff
@@ -1,68 +1,235 @@
...
+create an example project to show an integration with Gradle Build Scans.
+
+== What’s new and noteworthy
+
+As Spring Boot uses the Spring Framework 5.x the minimum Java version was
```
Comment (eriwen) [inline contents/index.adoc]: comma after 5.x is recommended

File: contents/index.adoc [hunk 7]
```diff
@@ -1,68 +1,235 @@
...
-  <body>
-    <script src="build/js/bundle.js"></script> // <1>
-  </body>
-</html>
+include::{samplescodedir}/sample-project/build.gradle[tags=plugins]
 ----
-<1> Webpack-generated JS bundle will include index.js and lodash
 
-== Step 2: TODO
+We now applied Gradle’s `java` and `application` plugin and as well added the
```
Comment (eriwen) [inline contents/index.adoc]: applied => apply?

File: contents/index.adoc [hunk 8]
```diff
@@ -1,68 +1,235 @@
...
+Gradle build scan plugin, with which we are able to create a build scan. We
+will cover what build scans are later in this section.
```
Comment (eriwen) [inline contents/index.adoc]: "section" => guide?

File: contents/index.adoc [hunk 9]
```diff
@@ -1,68 +1,235 @@
...
 
-Command-line tools can be invoked through Gradle https://docs.gradle.org/current/dsl/org.gradle.api.tasks.Exec.html[Exec] tasks.
+Next we need to add the dependencies needed to compile and run our example as
+we are not using Spring’s dependency management plugin. For that we use the
+newly introduced Gradle BOM support (still as feature preview) and load the
```
Comment (eriwen) [inline contents/index.adoc]: "the newly introduced Gradle BOM support (still as feature preview)" =>"Gradle's BOM support"?

File: contents/index.adoc [hunk 10]
```diff
@@ -1,68 +1,235 @@
...
+Spring Boot BOM file to be able to resolve all required dependencies with the
+proper version.
 
-NOTE: https://guides.gradle.org/writing-gradle-tasks/[Custom Task Classes] are highly recommended for any applications where configuration or reusability is desired; for example, to allow dev (debug with sourcemaps) and production variants.
+NOTE: If you’d like to read more about Gradle’s BOM support please visit
+https://docs.gradle.org/4.6/release-notes.html#bom-import[this page].
```
Comment (eriwen) [inline contents/index.adoc]: I found this doc: https://docs.gradle.org/current/userguide/managing_transitive_dependencies.html#sec:bom_import — is this more helpful than the release notes?

File: contents/index.adoc [hunk 11]
```diff
@@ -1,68 +1,235 @@
...
-    commandLine "$projectDir/node_modules/.bin/webpack", "app/index.js", "$buildDir/js/bundle.js"
-}
+include::{samplescodedir}/sample-project/build.gradle[tags=dependencies]
+----
+
+To comply with the Spring Boot BOM the `components` block is needed to
+strictly use the `snakeyaml` dependency with version `1.19` as the
+`spring-beans` dependency has version `1.20` as transitive dependency.
+
+As Gradle’s BOM support is still a feature preview, we need to enable this
```
Comment (eriwen) [inline contents/index.adoc]: Maybe say something like: "If using a Gradle version before 5.0, ..."

File: contents/index.adoc [hunk 12]
```diff
@@ -1,68 +1,235 @@
...
+`build.gradle` file. This will always publish a build scan after each build
+and always accept the https://gradle.com/legal/terms-of-service[license agreement].
+
+.build.gradle
+[source,groovy]
+----
+include::{samplescodedir}/sample-project/build.gradle[tags=buildScan]
+----
+
+== Creating a Hello Gradle sample application
```
Comment (eriwen) [inline contents/index.adoc]: Hello Gradle could optionally be quoted

File: contents/index.adoc [hunk 13]
```diff
@@ -1,68 +1,235 @@
...
+----
+./gradlew bootJar
+----
 
+The executable jar is located in the `build/libs` directory and you can run
+it by executing the following command:
+
+[listing]
+----
+java - jar build/libs/gradle-spring-boot-project.jar
```
Comment (eriwen) [inline contents/index.adoc]: Extra space before "jar"?

File: samples/code/sample-project/build.gradle [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+*/
+
+// Apply the build scan and Spring Boot 2.0.x plugin
+
+// tag::plugins[]
+plugins {
+    id 'java'
+    id 'application'
+    id 'com.gradle.build-scan' version '1.15.1'
+	id 'org.springframework.boot' version '2.0.3.RELEASE'
```
Comment (eriwen) [inline samples/code/sample-project/build.gradle]: Nitpick whitespace

File: samples/code/sample-project/sample-project.sample.out [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+> Task :compileTestJava
+> Task :processTestResources NO-SOURCE
+> Task :testClasses
+> Task :test
+> Task :check
+> Task :build
+
+10 actionable tasks: 10 executed
+
+Publishing build scan...
```
Comment (eriwen) [inline samples/code/sample-project/sample-project.sample.out]: This won't be there unless we run with `--scan` right?

File: samples/code/sample-project/sample-project.sample.conf [hunk 16]
```diff
@@ -0,0 +1,6 @@
+executable: gradle
+args: build
+flags: "--console verbose"
+expected-output-file: sample-project.sample.out
+allow-additional-output: true
+allow-disordered-output: true
```
Comment (eriwen) [inline samples/code/sample-project/sample-project.sample.conf]: Why is this needed?

File: build.gradle.kts (hunk 1 above)
Comment (guylabs) [reply build.gradle.kts]: Ok thanks for that. Maybe we should adapt the snippet here as well: https://github.com/gradle/exemplar/blob/master/README.md Should I create a PR for that?

File: contents/index.adoc (hunk 2 above)
Comment (guylabs) [reply contents/index.adoc]: done.

File: contents/index.adoc (hunk 3 above)
Comment (guylabs) [reply contents/index.adoc]: consistency is important. Therefore I did it ;)

File: contents/index.adoc (hunk 4 above)
Comment (guylabs) [reply contents/index.adoc]: Good point. Added it.

File: contents/index.adoc (hunk 5 above)
Comment (guylabs) [reply contents/index.adoc]: Fixed.

File: contents/index.adoc (hunk 6 above)
Comment (guylabs) [reply contents/index.adoc]: Fixed.

File: contents/index.adoc (hunk 7 above)
Comment (guylabs) [reply contents/index.adoc]: Fixed.

File: contents/index.adoc (hunk 8 above)
Comment (guylabs) [reply contents/index.adoc]: Fixed.

File: contents/index.adoc (hunk 9 above)
Comment (guylabs) [reply contents/index.adoc]: Good point! Fixed.

File: contents/index.adoc (hunk 10 above)
Comment (guylabs) [reply contents/index.adoc]: Yes this is enough and better to reference the user guide. Changed it.

File: contents/index.adoc (hunk 11 above)
Comment (guylabs) [reply contents/index.adoc]: Fixed.

File: contents/index.adoc (hunk 12 above)
Comment (guylabs) [reply contents/index.adoc]: Fixed.

File: contents/index.adoc (hunk 13 above)
Comment (guylabs) [reply contents/index.adoc]: Fixed.

File: samples/code/sample-project/build.gradle (hunk 14 above)
Comment (guylabs) [reply samples/code/sample-project/build.gradle]: Nice catch. Fixed.

File: samples/code/sample-project/sample-project.sample.out (hunk 15 above)
Comment (guylabs) [reply samples/code/sample-project/sample-project.sample.out]: Well the build scan plugin is configured with the `publishAlways()` option. Therefore it will be shown.

File: samples/code/sample-project/sample-project.sample.conf (hunk 16 above)
Comment (guylabs) [reply samples/code/sample-project/sample-project.sample.conf]: Well if you run it without the `allow-disordered-output` option the test fails, as the `test` and `check` task have a preceding empty line and we do not assert the test output of Sprint Boot.

Comment (eriwen) [reply samples/code/sample-project/sample-project.sample.conf]: Got it
